Automattic-Tracks-iOSTests/CrashLoggingTests.swift

```diff
@@ -0,0 +1,141 @@
+            expectation.fulfill()
+        }
+
+        CrashLogging.start(withDataProvider: mockDataProvider)
+
+        wait(for: [expectation], timeout: 1)
+    }
+
+    func testThatUserDataIsBeingStoredForUseInCrashLogs() {
+        mockDataProvider._currentUser = testUser
+        CrashLogging.start(withDataProvider: mockDataProvider)
+
+        XCTAssert(CrashLogging.sharedInstance.currentUser.email == testUser.email)
+        XCTAssert(CrashLogging.sharedInstance.cachedUser?.email == testUser.email)
+    }
+
+    func testThatEventsAreNotSentWhenUserOptsOut() {
+        mockDataProvider.userHasOptedOut = true
+
+        let expectation = XCTestExpectation(description: "Events should not be sent if user has opted out")
+
+        CrashLogging.sharedInstance.shouldSendEventCallback = { shouldSendEvent in
+            expectation.isInverted = shouldSendEvent   // fail if `shouldSendEvent` is true
+            expectation.fulfill()
+        }
+
+        CrashLogging.start(withDataProvider: mockDataProvider)
+        CrashLogging.logMessage("This is a test")
+
+        wait(for: [expectation], timeout: 1)
+    }
+
+    func testThatEventsAreSentWhenUserOptsIn() {
+        mockDataProvider._currentUser = testUser
+
+        let expectation = XCTestExpectation(description: "Events be sent if user has opted in")
+
+        CrashLogging.sharedInstance.shouldSendEventCallback = { shouldSendEvent in
+            expectation.isInverted = !shouldSendEvent   // succeed if `shouldSendEvent` is true
+            expectation.fulfill()
+        }
+
+        CrashLogging.start(withDataProvider: mockDataProvider)
+        CrashLogging.logMessage("This is a test")
+
+        wait(for: [expectation], timeout: 1)
+    }
+
+//    func testThatLoggedErrorsContainAStackTrace() {
+//        let expectation = XCTestExpectation(description: "Event should contain a Stack Trace")
+//
+//        mockDataProvider.didLogErrorCallback = { event in
+//            expectation.isInverted = event.stacktrace == nil   // fail if `stacktrace` is nil
+//            expectation.fulfill()
+//        }
+//
+//        CrashLogging.start(withDataProvider: mockDataProvider)
+//        CrashLogging.logError("This is a test")
+//
+//        wait(for: [expectation], timeout: 1)
+//    }
+//
+//    func testThatLoggedMessagesContainAStackTrace() {
+//        let expectation = XCTestExpectation(description: "Event should contain a Stack Trace")
+//
+//        mockDataProvider.didLogMessageCallback = { event in
+//            debugPrint("=== \(String(describing: event.stacktrace))")
+//            expectation.isInverted = event.stacktrace == nil   // fail if `stacktrace` is nil
+//            expectation.fulfill()
+//        }
+//
+//        CrashLogging.start(withDataProvider: mockDataProvider)
+//        CrashLogging.logMessage("This is a test")
+//
+//        wait(for: [expectation], timeout: 1)
+//    }
+}
+
+/// Allow throwing Strings as error
+extension String: Error {}
```

Comment:
It's used as a shortcut here: https://github.com/Automattic/Automattic-Tracks-iOS/pull/103/files/8d80f1f346eb45f73f7f8081284c279cee2a5933#diff-54236c2e7c3ce4c8904239c3cdc9100fR94 to avoid creating an entire `Error` object to test with! 😄

Reply:
The problem with this is the effect is apparently global. All `String` objects will now conform to `Error`, not just in the test. I'm not sure if we want that.

Reply:
I agree that that would be bad! But it seems to be confined to the test target?

Unless I'm missing something, which is 100% possible! :) 

<img width="2118" alt="Screen Shot 2019-05-29 at 9 49 55 AM" src="https://user-images.githubusercontent.com/1123407/58571583-48374e00-81f7-11e9-94de-e38f9ebbf831.png">

Reply:
Oh, nice. I was wrong. 👍